sal/tests/client/unit/test_host.py

```diff
@@ -0,0 +1,148 @@
+"""
+Tests setting of the host on the client
+
+.. note::
+    These tests assume that the SALClient connects to the host when it is
+    initialised. Although it would be preferable to test host connection using
+    SALClient.host, there is no straightforward way to initialise attributes
+    without connecting to the host.  
+"""
+
+from unittest.mock import patch
+
+import requests
+import pytest
+
+from sal.client.main import SALClient, exception
+
+
+def test_connect_to_valid_host(host, server_root_response):
+
+    """
+    GIVEN
+        A URL to a valid SAL server
+    WHEN
+        A client connects to the host
+    THEN
+        The client validates the host (i.e. no exceptions are thrown)
+        AND the client stores the host URL
+        AND the client determines if the host has authentication enabled
+    """
+
+    requires_auth = server_root_response.json()['api']['requires_auth']
+
+    with patch('sal.client.main.requests.get',
+                return_value=server_root_response):
+        sc = SALClient(host)
+        assert sc.host == host
+        assert sc.auth_required == requires_auth
+
+
+
+@pytest.mark.xfail(reason='AttributeError when SALClient._host undefined',
+                   raises=AttributeError)
+def test_connect_to_host_no_response():
+
+    """
+    GIVEN
+        A URL to which the client cannot connect
+    WHEN
+        A client attempts to connect to the host
+    THEN
+        The client raises a ConnectionError
+    """
+
+    with patch('sal.client.main.requests.get',
+               side_effect=requests.ConnectionError()):
+        with pytest.raises(ConnectionError):
+            SALClient('https://sal.testing')
+
+
+
+def test_connect_to_host_not_sal_server(server_root_response):
+
+    """
+    GIVEN
+        A URL to which the client can connect, but the host is not a SAL server
+    WHEN
+        A client connects to the host
+    THEN
+        The client raises a ConnectionError
+    """
+
+    server_root_response.json.return_value['service']['name'] = 'Generic Server'
+
+    with patch('sal.client.main.requests.get',
+               return_value=server_root_response):
+        with pytest.raises(ConnectionError):
+            SALClient('https://sal.testing')
+
+
+def test_connect_to_host_with_new_api(server_root_response):
+
+    """
+    GIVEN
+        A URL for a SAL server with a newer API than the client
+    WHEN
+        The client connects to the host
+    THEN
+        The client raises a ConnectionError
+    """
+
+    server_root_response.json.return_value['api']['version'] = 3
```

Comment:
**suggeston:** Instead of hard coding the future API version, would it be possible to get the current version and just add one?

Reply:
Good idea, I've implemented this.